gtsam_unstable/geometry/Similarity3.cpp

```diff
@@ -85,10 +85,110 @@ Point3 Similarity3::transformFrom(const Point3& p, //
   return s_ * q;
 }
 
+Pose3 Similarity3::transformFrom(const Pose3& T) const {
+  Rot3 R = R_.compose(T.rotation());
+  Point3 t = Point3(s_ * (R_.matrix() * T.translation().vector() + t_.vector()));
+  return Pose3(R, t);
+}
+
 Point3 Similarity3::operator*(const Point3& p) const {
   return transformFrom(p);
 }
 
+// Refer to: http://www5.informatik.uni-erlangen.de/Forschung/Publikationen/2005/Zinsser05-PSR.pdf Chapter 3
+Similarity3 Similarity3::Align(const std::vector<Point3Pair>& abPointPairs) {
+  const size_t n = abPointPairs.size();
+  if (n < 3) throw std::runtime_error("input should have at least 3 pairs of points");  // we need at least three pairs
+
+  // calculate centroids
+  Point3 aCentroid(0, 0, 0), bCentroid(0, 0, 0);
+  for (const Point3Pair& abPair : abPointPairs) {
```

Comment:
From https://stackoverflow.com/questions/21300410/using-stdtie-as-a-range-for-loop-target:

```
Point3 a, b;
for (auto& p : abPointPairs) {
  std::tie(a, b) = std::move(p);
  // do stuff;
}
```